### ISSUE TITLE
Fix CLI `account_update` related commands

### ISSUE THREAD

**Issue by abitmore** [bitshares/bitshares-core#576]
The account update commands in CLI are known to be broken.
* set_voting_proxy
* update_worker_votes
* vote_for_committee_member
* vote_for_witness

They're useful, so we should keep & fix them.

//Update with more details:
Specifically, the commands are setting wrong data in `new_options` field. For example, if you vote for a witness with a new account, it's fine; if you then vote for another witness or for a committee member, the previously voted one will be dropped.

**Comment by jmjatlanta** [comment]
I took a look at this, and created a test case. When I attempt a "set_voting_proxy" i get:
```
1582835ms th_a       wallet.cpp:1857               sign_transaction     ] Caught exception while broadcasting tx 369e6a67fbb54c56f20c84a160fd96a23c5722ae:  0 exception: unspecified
missing required active authority: Missing Active Authority 1.2.18
    {"error":"missing required active authority: Missing Active Authority 1.2.18","data":{"id":44,"jsonrpc":"2.0","error":{"code":1,"message":"missing required active authority: Missing Active Authority 1.2.18","data":{"code":3030001,"name":"tx_missing_active_auth","message":"missing required active authority","stack":[{"context":{"level":"error","file":"transaction.cpp","line":284,"method":"verify_authority","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"Missing Active Authority ${id}","data":{"id":"1.2.18","auth":{"weight_threshold":1,"account_auths":[],"key_auths":[["BTS6iAeb6TJg3A8hgeNvRj1pdBoFvgzLpTGchs63T64Jwn86jBPFR",1]],"address_auths":[]},"owner":{"weight_threshold":1,"account_auths":[],"key_auths":[["BTS7eE7iVv5LD7VEDPacL4UE4k4FMj8vNWZnK99B2yFCSJ3pTkyJS",1]],"address_auths":[]}}},{"context":{"level":"warn","file":"transaction.cpp","line":299,"method":"verify_authority","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"ops":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"sigs":[]}},{"context":{"level":"warn","file":"transaction.cpp","line":387,"method":"verify_authority","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"*this":{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}}},{"context":{"level":"warn","file":"db_block.cpp","line":628,"method":"_apply_transaction","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"trx":{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}}},{"context":{"level":"warn","file":"db_block.cpp","line":227,"method":"push_transaction","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"trx":{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}}},{"context":{"level":"warn","file":"websocket_api.cpp","line":122,"method":"on_message","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"call.method":"call","call.params":[3,"broadcast_transaction",[{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}]]}}]}}}}
    th_a  state.cpp:38 handle_reply
1582838ms th_a       database_api.cpp:266          ~database_api_impl   ] freeing database api 4666691232
1582839ms th_a       database_api.cpp:266          ~database_api_impl   ] freeing database api 4666690400
1582841ms th_a       main.cpp:227                  test_method          ] e.to_detail_string(): 0 exception: unspecified
missing required active authority: Missing Active Authority 1.2.18
    {"error":"missing required active authority: Missing Active Authority 1.2.18","data":{"id":44,"jsonrpc":"2.0","error":{"code":1,"message":"missing required active authority: Missing Active Authority 1.2.18","data":{"code":3030001,"name":"tx_missing_active_auth","message":"missing required active authority","stack":[{"context":{"level":"error","file":"transaction.cpp","line":284,"method":"verify_authority","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"Missing Active Authority ${id}","data":{"id":"1.2.18","auth":{"weight_threshold":1,"account_auths":[],"key_auths":[["BTS6iAeb6TJg3A8hgeNvRj1pdBoFvgzLpTGchs63T64Jwn86jBPFR",1]],"address_auths":[]},"owner":{"weight_threshold":1,"account_auths":[],"key_auths":[["BTS7eE7iVv5LD7VEDPacL4UE4k4FMj8vNWZnK99B2yFCSJ3pTkyJS",1]],"address_auths":[]}}},{"context":{"level":"warn","file":"transaction.cpp","line":299,"method":"verify_authority","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"ops":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"sigs":[]}},{"context":{"level":"warn","file":"transaction.cpp","line":387,"method":"verify_authority","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"*this":{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}}},{"context":{"level":"warn","file":"db_block.cpp","line":628,"method":"_apply_transaction","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"trx":{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}}},{"context":{"level":"warn","file":"db_block.cpp","line":227,"method":"push_transaction","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"trx":{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}}},{"context":{"level":"warn","file":"websocket_api.cpp","line":122,"method":"on_message","hostname":"","thread_name":"th_a","timestamp":"2018-02-15T21:26:22"},"format":"","data":{"call.method":"call","call.params":[3,"broadcast_transaction",[{"ref_block_num":0,"ref_block_prefix":0,"expiration":"2018-02-15T21:26:50","operations":[[6,{"fee":{"amount":2005273,"asset_id":"1.3.0"},"account":"1.2.18","new_options":{"memo_key":"BTS6dtXpcVyDvtpfEkk827fAPHpNGNeC822comrKVHWGKa3z6R6xy","voting_account":"1.2.17","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"extensions":{}}]],"extensions":[],"signatures":[]}]]}}]}}}}
    th_a  state.cpp:38 handle_reply

    {"account_to_modify":"jmjatlanta","voting_account":"nathan","broadcast":true}
    th_a  wallet.cpp:1775 set_voting_proxy 
```
Is that what I should get, or is there an error in my test?

**Comment by oxarbitrage** [comment]
yes, thats the problem.

**Comment by abitmore** [comment]
@jmjatlanta it's another issue. You need the private keys to be able to update an account, that's by design.

**Comment by xeroc** [comment]
Witness votes are tallied once every maintenance interval. Thats probably why you dont see the votes yet. As for the other error, could caching be the reason?

**Comment by abitmore** [comment]
Specifications.

When voting with a proxy, if then vote for a witness, what should we expect on the chain? Then, what should we expect from CLI?

Tallying is one issue here, but won't fix so far.

Caching should have been fixed.

There are still other bugs in the commands. Described in updated OP.

**Comment by xeroc** [comment]
You can add to your list account_upgrade for LTM.
With a new cli-wallet, you cannot import an accou t, upgrade it to LTM and start creating new accounts right away because the account caching still thinks the account is non-LTM

**Comment by abitmore** [comment]
This ticket is not for the caching issue. By the way I thought the caching issue has been fixed/got around by https://github.com/bitshares/bitshares-core/pull/640. @xeroc can you check? If not, we can create a new ticket for it, or add it to https://github.com/bitshares/bitshares-core/issues/151..

**Comment by jmjatlanta** [comment]
Caching: I'm not sure if it is a good test, but in my testnet, I am verifying that upgrade_account immediately changes the membership_expiration_date. Also, is_lifetime_member() returns true. So it looks like those two areas work as expected.

vote_for_witness: In my test, I am calling vote_for_witness("my_account", "init1", true, true). Afterwards, I am calling get_witness, and comparing the total_votes from what it was before the call. It doesn't seem to increase witness_object.total_votes (0 == 0). Does "get_witness("init1").total_votes" need to wait for a maintenance interval to be updated?

**Comment by abitmore** [comment]
> I then vote for init2, and receive the following transaction back:
>```
>{
> "id": "1.6.2",
>...
>}

This is not the result of "vote_for..." command (not a transaction json), but perhaps is result of following "get_witness" command. If showed results of "vote_for..." commands I may clearly know whether the bug mentioned in OP is fixed, although from another result of `get_witness` command below I guess it's fixed.

**Comment by abitmore** [comment]
Confirmed that it's caused by the caching issue, should have been fixed by #640. Closing this ticket.

@jmjatlanta I'd like to merge the skeleton of your unit test code (for testing CLI) into this repo, but it need some cleanup and maybe abstraction and refactory. I'll create another ticket for this. Thanks your your contribution.
